### ISSUE TITLE
Test failures in testing/eunit

### ISSUE THREAD

**Issue by joshuabach** [indigo-dc/tts#418]
Some of the test in branch testing/eunit, fail and I'm not sure why.

Please also take al look at [test/watts_test.erl:64](https://github.com/indigo-dc/tts/blob/testing/eunit/test/watts_test.erl#L64), as I'm not sure how to test this properly.

Otherwise coverage is >70% in all test modules now.

**Comment by bwegh** [comment]
which are failing for you?
for me the test
' watts_rest_test:get_json_test/0' is failing with
```
 Failure/Error: {error,
                        {badmap,<<"CredId">>},
                        [{maps,with,
                             [[cred_id,ctime,interface,service_id],
                              <<"CredId">>],
                             [{file,"maps.erl"},{line,230}]},
                         {watts_rest,'-return_json_credential_list/2-fun-0-',
                             3,
                             [{file,
                                  "/home/bas/Documents/s/sw/tts/_build/test/lib/watts/src/watts_rest.erl"},
                              {line,366}]},
                         {lists,foldl,3,[{file,"lists.erl"},{line,1262}]},
                         {watts_rest,return_json_credential_list,2,
                             [{file,
                                  "/home/bas/Documents/s/sw/tts/_build/test/lib/watts/src/watts_rest.erl"},
                              {line,374}]},
                         {watts_rest,get_json,2,
                             [{file,
                                  "/home/bas/Documents/s/sw/tts/_build/test/lib/watts/src/watts_rest.erl"},
                              {line,220}]},
                         {watts_rest_test,'-get_json_test/0-fun-1-',2,
                             [{file,
                                  "/home/bas/Documents/s/sw/tts/_build/test/lib/watts/test/watts_rest_test.erl"},
                              {line,345}]},
                         {lists,foldl,3,[{file,"lists.erl"},{line,1262}]},
                         {watts_rest_test,get_json_test,0,
                             [{file,
                                  "/home/bas/Documents/s/sw/tts/_build/test/lib/watts/test/watts_rest_test.erl"},
                              {line,349}]}]}
```
on 
```
Expecting #{<<"credential_list">> => [<<"Cred1">>]} on state {state,get,
                                                              latest,
                                                              credential,
                                                              undefined,
                                                              undefined,
                                                              undefined,
                                                              undefined,pid1}
```

**Comment by joshuabach** [comment]
Exactly that one. In line 366 in watts_rest.erl `maps:with` gets passed `<<"CredId">>`, which is not a map. (Test fails even before the assertion is tried).

**Comment by bwegh** [comment]
this helps:
```
diff --git a/test/watts_rest_test.erl b/test/watts_rest_test.erl
index c25a237..85e6dde 100644
--- a/test/watts_rest_test.erl
+++ b/test/watts_rest_test.erl
@@ -336,7 +336,10 @@ get_json_test() ->
                             id = undefined,
                             session_pid = pid1,
                             method = get},
-                     #{<<"credential_list">> => [<<"Cred1">>]}
+                     #{<<"credential_list">> => [#{<<"cred_id">> => <<"CRED1">>,
+                                              <<"ctime">> => <<"now">>,
+                                              <<"interface">> => <<"test">>,
+                                              <<"service_id">> => <<"test1">>}]}
                     }
                    ],
 
@@ -508,7 +511,12 @@ start_meck() ->
                   end,
     GetCredList = fun(SessionPid) ->
                           case SessionPid of
-                              pid1 -> {ok, [<<"CredId">>]};
+                              pid1 -> {ok, [#{cred_id => <<"CRED1">>,
+                                              ctime => <<"now">>,
+                                              interface => <<"test">>,
+                                              service_id => <<"test1">>,
+                                              removed => true
+                                             }]};
                               _ -> {ok, []}
                           end
               end,
```

**Comment by bwegh** [comment]
regarding watts_test:
```
diff --git a/test/watts_test.erl b/test/watts_test.erl
index ed4b495..626a198 100644
--- a/test/watts_test.erl
+++ b/test/watts_test.erl
@@ -3,6 +3,7 @@
 
 
 -define(ISSUER_URL, <<"https:://issuer.sample">>).
+-define(ISSUER2_URL, <<"https:://issuer.other">>).
 
 error_session_test() ->
     {ok, Meck} = start_meck(),
@@ -44,10 +45,10 @@ login_and_out_test() ->
     ok.
 
 additional_login_test() ->
-    %% TODO This test probably requires unmecked session mgmt to be meaningful
     {ok, Meck} = start_meck(),
     try
         Issuer = ?ISSUER_URL,
+        Issuer2 = ?ISSUER2_URL,
         OidcToken1 = #{id => #{claims => #{sub => <<"sub">>, iss => Issuer}},
                        access => #{token => <<"accestoken">>},
                        cookies => []
@@ -55,13 +56,17 @@ additional_login_test() ->
         {ok, #{session_pid := Pid}} = watts:login_with_oidcc(OidcToken1),
 
         {ok, Token} = watts_session:get_sess_token(Pid),
-        OidcToken2 = #{id => #{claims => #{sub => <<"othersub">>, iss => Issuer}},
+        OidcToken2 = #{id => #{claims => #{sub => <<"sub">>, iss => Issuer2}},
                        access => #{token => <<"accestoken">>},
                        cookies => [{watts_http_util:cookie_name(), Token}]
                       },
+        Params = [],
+        ServiceId = <<"service0815">>,
+        Provider = <<"ID3">>,
+        ok = watts_session:set_redirection(ServiceId, Params, Provider, Pid),
         {ok, #{session_pid := Pid}} = watts:login_with_oidcc(OidcToken2),
-
-        %%TODO true = watts_userinfo:has_additional_login(_, _, watts_session:get_user_info(Pid)),
+        {ok, UserInfo} = watts_session:get_user_info(Pid),
+        true = watts_userinfo:has_additional_login(ServiceId, Provider, UserInfo),
 
         ok = watts:logout(Pid),
         test_util:wait_for_process_to_die(Pid, 100)
@@ -152,7 +157,8 @@ start_meck() ->
     ProviderList = fun() ->
                              {ok, [ {<<"id">>, pid1},
                                     {<<"ID1">>, pid2},
-                                    {<<"ID2">>, pid3}]}
+                                    {<<"ID2">>, pid3},
+                                    {<<"ID3">>, pid4}]}
                      end,
 
     RetrieveUserInfo = fun(_, _, _) ->
@@ -165,6 +171,13 @@ start_meck() ->
                                 ready => true,
                                 extra_config => #{priority => 5},
                                 issuer => ?ISSUER_URL}};
+                      (<<"ID3">>) ->
+                           {ok, #{
+                                id => <<"ID3">>,
+                                description => <<"a second provider">>,
+                                ready => true,
+                                extra_config => #{priority => 5},
+                                issuer => ?ISSUER2_URL}};
                       (_) ->
                              {ok, #{
                                 id => <<"ID1">>,
@@ -173,9 +186,13 @@ start_meck() ->
                                 extra_config => #{priority => 5},
                                 issuer => ?ISSUER_URL}}
                      end,
-    FindProvider = fun(_Issuer) ->
-                           {ok, provider_pid}
-                     end,
+    FindProvider = fun(Issuer) ->
+                           case Issuer of
+                               ?ISSUER_URL -> {ok, provider_pid};
+                               ?ISSUER2_URL -> {ok, <<"ID3">>};
+                               _ -> {error, not_found}
+                           end
+                   end,
     NewSession = fun() ->
                          TestSession
                  end,
```

**Comment by bwegh** [comment]
fixed with #422
